tests/OutputTest/OutputTest.php

```diff
@@ -0,0 +1,209 @@
+        $flags = FilesystemIterator::KEY_AS_FILENAME
+            | FilesystemIterator::CURRENT_AS_FILEINFO
+            | FilesystemIterator::SKIP_DOTS;
+        $filter = function (SplFileInfo $file) {
+            return $file->getExtension() === "html";
+        };
+        $dir = new RecursiveDirectoryIterator(self::DATASET_DIRECTORY, $flags);
+        $files = new CallbackFilterIterator(new RecursiveIteratorIterator($dir), $filter);
+
+        foreach ($files as $file) {
+            $name = self::datasetName($file);
+            yield new Dataset($name, $file);
+        }
+    }
+
+    public static function outputTestProvider(): Iterator
+    {
+        foreach (self::datasets() as $dataset) {
+            yield $dataset->name => [$dataset];
+        }
+    }
+
+    protected function setUp(): void
+    {
+        $process = new Process(["gs", "-v"]);
+        $exitCode = $process->run();
+
+        if ($exitCode === 127) {
+            $this->markTestSkipped("Output tests need Ghostscript to be installed");
+        }
+    }
+
+    /**
+     * @dataProvider outputTestProvider
+     */
+    public function testOutputMatchesReferenceRendering(Dataset $dataset): void
+    {
+        $document = $dataset->render();
+        $referenceFile = $dataset->referenceFile()->getPathname();
+        $actualOutputFile = tempnam(sys_get_temp_dir(), "dompdf_test_");
+
+        file_put_contents($actualOutputFile, $document->output());
+
+        try {
+            $this->assertOutputMatches($referenceFile, $actualOutputFile);
+        } catch (AssertionFailedError $e) {
+            $this->saveFailedOutput($dataset);
+            throw $e;
+        } finally {
+            unlink($actualOutputFile);
+        }
+    }
+
+    private function assertOutputMatches(
+        string $referenceFile,
+        string $actualOutputFile
+    ): void {
+        $command = function ($file) {
+            return [
+                "gs",
+                "-q", "-dBATCH", "-dNOPAUSE", "-sstdout=%stderr",
+                "-sDEVICE=png16m", "-dGraphicsAlphaBits=4",
+                "-sOutputFile=-", $file
+            ];
+        };
+        $process1 = new Process($command($referenceFile));
+        $process2 = new Process($command($actualOutputFile));
+
+        $process1->mustRun();
+        $process2->mustRun();
+
+        $error1 = $process1->getErrorOutput();
+        $error2 = $process2->getErrorOutput();
+
+        if ($error1 !== "") {
+            throw new RuntimeException("Ghostscript execution failed: $error1");
+        }
+
+        if ($error2 !== "") {
+            throw new RuntimeException("Ghostscript execution failed: $error2");
+        }
+
+        $referenceImages = $this->outputToImageData($process1->getOutput());
+        $actualImages = $this->outputToImageData($process2->getOutput());
+
+        $expectedCount = count($referenceImages);
+        $failureMessage = "Output does not match reference rendering (number of pages is different).";
```

Comment:
```suggestion
        $failureMessage = sprintf('Output (count: `%d`) does not match reference (count: `%d`) rendering (number of pages is different).', $expectedCount, $actualImages);
```

It would be nice to have the numbers in the failure message just like in the default error message for `assertCount`

Also should we not always use single quote `'` instead of double `"` or are there no styling guidelines

Reply:
I think I left them out because the numbers are contained in the default error message, which is also displayed. But I think you are right, this is nicer. Included you suggestion with a slightly different wording.

Reply:
> Also should we not always use single quote `'` instead of double `"` or are there no styling guidelines

Other than the `phpcs.xml` configuration, there are no style guidelines that I am aware of. The code base uses double quotes predominantly, though (apart from the `Cpdf` class), so that is what I have been sticking to.

Reply:
Yeah, no real styling guide. I think the majority of changes in recent years has leaned towards double quotes. I believe primarily due to the decrease in performance differential between the two and to ensure more consistent code presentation.